source/extensions/filters/http/cache/http_cache_utils.cc

```diff
@@ -0,0 +1,175 @@
+  // quoted-string   = DQUOTE *( qdtext / quoted-pair ) DQUOTE
+  // qdtext          = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
+  // obs-text        = %x80-FF
+  // quoted-pair     = "\" ( HTAB / SP / VCHAR / obs-text )
+  // VCHAR           =  %x21-7E  ; visible (printing) characters
+  Duration max_age = -InfiniteDuration();
+  bool found_s_maxage = false;
+  while (!cache_control.empty()) {
+    // Each time through the loop, we eat one cache-directive. Each branch
+    // either returns or completely eats a cache-directive.
+    if (ConsumePrefix(&cache_control, "no-cache")) {
+      if (eatToken(cache_control)) {
+        // The token wasn't no-cache; it just started that way, so we must
+        // finish eating this cache-directive.
+        if (ConsumePrefix(&cache_control, "=")) {
+          eatDirectiveArgument(cache_control);
+        }
+      } else {
+        // Found a no-cache directive, so validation is required.
+        return -InfiniteDuration();
+      }
+    } else if (ConsumePrefix(&cache_control, "s-maxage=")) {
+      max_age = eatLeadingDuration(&cache_control);
+      found_s_maxage = true;
+      cache_control = StripLeadingAsciiWhitespace(cache_control);
+      if (!cache_control.empty() && cache_control[0] != ',') {
+        // Unexpected text at end of directive
+        return -InfiniteDuration();
+      }
+    } else if (!found_s_maxage && ConsumePrefix(&cache_control, "max-age=")) {
+      max_age = eatLeadingDuration(&cache_control);
+      if (!cache_control.empty() && cache_control[0] != ',') {
+        // Unexpected text at end of directive
+        return -InfiniteDuration();
+      }
+    } else if (eatToken(cache_control)) {
+      // Unknown directive--ignore.
+      if (ConsumePrefix(&cache_control, "=")) {
+        eatDirectiveArgument(cache_control);
+      }
+    } else {
+      // This directive starts with illegal characters. Require validation.
+      return -InfiniteDuration();
+    }
+    // Whichever branch we took should have consumed the entire cache-directive,
+    // so we just need to eat the delimiter and optional whitespace.
+    ConsumePrefix(&cache_control, ",");
+    cache_control = StripLeadingAsciiWhitespace(cache_control);
+  }
+  return max_age;
+}
+
+Time httpTime(const HeaderEntry* header_entry) {
+  if (!header_entry) {
+    return InfinitePast();
+  }
+  Time time;
+  const string input(header_entry->value().getStringView());
+
+  // RFC 7231 7.1.1.1: Acceptable Date/Time Formats:
+  // Sun, 06 Nov 1994 08:49:37 GMT    ; IMF-fixdate
+  // Sunday, 06-Nov-94 08:49:37 GMT   ; obsolete RFC 850 format
+  // Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
+  const array<string, 3> rfc7231_date_formats = {
+      "%a, %d %b %Y %H:%M:%S GMT", "%A, %d-%b-%y %H:%M:%S GMT", "%a %b %e %H:%M:%S %Y"};
+  for (const string& format : rfc7231_date_formats) {
+    if (ParseTime(format, input, &time, nullptr)) {
```

Comment:
since we're in Envoy should we return as std::chrono?

Reply:
Done.